modules/twinkleblock.js

```diff
@@ -0,0 +1,879 @@
+ ****************************************
+ * Mode of invocation:     Tab ("Block")
+ * Active on:              any page with relevant user name (userspace, contribs, etc.)
+ * Config directives in:   [soon to be TwinkleConfig]
+ */
+
+Twinkle.block = function twinkleblock() {
+  // should show on Contributions pages, anywhere there's a relevant user
+  if ( Morebits.userIsInGroup('sysop') && relevantUser ) {
+    Twinkle.addPortletLink(Twinkle.block.callback, "Block", "tw-block", "Block relevant user" );
+  }
+};
+
+// Contains the current block info in an object
+// Once filled, it will look something like:
+// { expiry: <some date>, anononly: true, ... }
+Twinkle.block.currentBlockInfo = undefined;
+
+Twinkle.block.callback = function twinkleblockCallback() {
+
+  var Window = new Morebits.simpleWindow( 650, 530 );
+  // need to be verbose about who we're blocking
+  Window.setTitle( "Block or template " + relevantUser );
+  Window.setScriptName( "Twinkle" );
+  Window.addFooterLink( "Block templates", "Template:Uw-block/doc/Block_templates" );
+  Window.addFooterLink( "Block policy", "WP:BLOCK" );
+  Window.addFooterLink( "Twinkle help", "WP:TW/DOC#block" );
+
+  formCleared = false; // used by evaluate()
+
+  var form = new Morebits.quickForm( Twinkle.block.callback.evaluate );
+  var actionfield = form.append( {
+      type: 'field',
+      label: 'Type of action'
+    } );
+  actionfield.append({
+      type: 'checkbox',
+      name: 'actiontype',
+      event: Twinkle.block.callback.changeAction,
+      list: [
+        {
+          label: 'Block user',
+          value: 'block',
+          tooltip: 'Block the relevant user with given options.',
+          checked: true
+        },
+        {
+          label: 'Add block template to user talk page',
+          value: 'template',
+          tooltip: 'If the blocking admin forgot to issue a block template, or you have just blocked the user without templating them, you can use this to issue the appropriate template.',
+          checked: true
+        }
+      ]
+    });
+
+  form.append({ type: 'field', label: 'Preset', name: 'field_preset' });
+  form.append({ type: 'field', label: 'Template options', name: 'field_template_options' });
+  form.append({ type: 'field', label: 'Block options', name: 'field_block_options' });
+
+  form.append( { type:'submit' } );
+
+  result = form.render();
+  Window.setContent( result );
+  Window.display();
+
+  Twinkle.block.fetchUserInfo(function() {
+    // init the controls after user and block info have been fetched
+    var evt = document.createEvent( "Event" );
+    evt.initEvent( 'change', true, true );
+    result.actiontype[0].dispatchEvent( evt );
+  });
+};
+
+Twinkle.block.fetchUserInfo = function twinkleblockFetchUserInfo(fn) {
+
+  api.get({
+    format: 'json',
+    action: 'query',
+    list: 'blocks|users',
+    bkusers: mw.config.get('wgRelevantUserName'),
+    ususers: mw.config.get('wgRelevantUserName')
+  })
+  .then(function(data){
```

Comment:
I'm not familiar with `then`. Does the callback still run even if the API query fails (e.g. network connection is down)? If so, this is going to cause JavaScript errors, because `data.query` will be undefined, and trying to access `data.query.blocks` will cause a ReferenceError.

I think this code should only run when the API query succeeds. Isn't there a `done` method, or similar?